// @iden3/js-iden3-auth/dist/node/cjs/circuits/atomicSigV2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AtomicQuerySigV2PubSignalsVerifier = void 0;
const query_1 = require("../circuits/query");
const ownershipVerifier_1 = require("../circuits/ownershipVerifier");
const common_1 = require("../circuits/common");
const js_iden3_core_1 = require("@iden3/js-iden3-core");
const js_sdk_1 = require("@0xpolygonid/js-sdk");
const constants_1 = require("../constants");
class AtomicQuerySigV2PubSignalsVerifier extends ownershipVerifier_1.IDOwnershipPubSignals {
    constructor(pubSignals) {
        super();
        this.pubSignals = new js_sdk_1.AtomicQuerySigV2PubSignals();
        this.pubSignals = this.pubSignals.pubSignalsUnmarshal(js_sdk_1.byteEncoder.encode(JSON.stringify(pubSignals)));
        this.userId = this.pubSignals.userID;
        this.challenge = this.pubSignals.requestID;
    }
    async verifyQuery(query, schemaLoader, verifiablePresentation, opts) {
        const outs = {
            issuerId: this.pubSignals.issuerID,
            schemaHash: this.pubSignals.claimSchema,
            slotIndex: this.pubSignals.slotIndex,
            operator: this.pubSignals.operator,
            value: this.pubSignals.value,
            timestamp: this.pubSignals.timestamp,
            merklized: this.pubSignals.merklized,
            claimPathKey: this.pubSignals.claimPathKey,
            claimPathNotExists: this.pubSignals.claimPathNotExists,
            valueArraySize: constants_1.CONSTANTS.CIRCUITS_ARRAY_VALUE_SIZE,
            isRevocationChecked: this.pubSignals.isRevocationChecked
        };
        await (0, query_1.checkQueryV2Circuits)(js_sdk_1.CircuitId.AtomicQuerySigV2, query, outs, schemaLoader, opts, verifiablePresentation);
        return this.pubSignals;
    }
    async verifyStates(resolvers, opts) {
        const resolver = (0, common_1.getResolverByID)(resolvers, this.pubSignals.issuerID);
        if (!resolver) {
            throw new Error(`resolver not found for issuerID ${this.pubSignals.issuerID.string()}`);
        }
        await (0, common_1.checkUserState)(resolver, this.pubSignals.issuerID, this.pubSignals.issuerAuthState);
        if (this.pubSignals.isRevocationChecked === 0) {
            return;
        }
        const issuerNonRevStateResolved = await (0, common_1.checkIssuerNonRevState)(resolver, this.pubSignals.issuerID, this.pubSignals.issuerClaimNonRevState);
        const acceptedStateTransitionDelay = opts?.acceptedStateTransitionDelay ?? constants_1.CONSTANTS.ACCEPTED_STATE_TRANSITION_DELAY;
        if (!issuerNonRevStateResolved.latest) {
            const timeDiff = Date.now() -
                (0, js_iden3_core_1.getDateFromUnixTimestamp)(Number(issuerNonRevStateResolved.transitionTimestamp)).getTime();
            if (timeDiff > acceptedStateTransitionDelay) {
                throw new Error('issuer state is outdated');
            }
        }
    }
}
exports.AtomicQuerySigV2PubSignalsVerifier = AtomicQuerySigV2PubSignalsVerifier;
